### ISSUE TITLE
12. Tagging Task

### ISSUE THREAD

**Issue by p-jiho** [p-jiho/natural-language#12]
- 태깅 작업 : 각 단어가 어떤 유형에 속해있는지를 알아내는 작업
- 개체명 인식 : 사람, 장소, 단체 등
- 품사 태깅 : 명사, 동사, 형용사 등

## 1. 케라스를 이용한 태깅 작업
- 개체명 인식기, 품사 태거의 공통점 : RNN의 다-대-다 작업, 양방향 RNN

#### 1) 훈련 데이터에 대한 이해
- X, y는 병렬 구조를 가져 둘의 데이터의 길이는 같다.

#### 2) Sequence Labeling
- 입력 시퀀스에 대하여 레이블 시퀀스를 부여하는 작업
- 태깅 작업은 대표적인 시퀀스 레이블 작업

#### 3) Bidirectional LSTM
- `model.add(Bidirectional(LSTM(hidden_units, return_sequences = True)))`

#### 4) RNN의 다-대-다 문제
- RNN의 은닉층은 `return_sequences`의 값을 통해 모든 시점을 출력할수도 마지막 시점만 출력할 수도 있다.
- 태깅 작업은 다-대-다 문제 : `return_sequences=True`로 설정해야한다.

<br></br>

#### 1) 품사 태깅 데이터에 대한 이해와 전처리
- `nltk.corpus.treebanck.tagged_sents()` : 영어 코퍼스에 토큰화, 품사 태깅 전처리를 진행한 문장 데이터
- 단어와 품사 태깅 정보 분리 : `zip()`
      - 각 문장 샘플을 `for` 문을 이용해 하나씩 가져온다.
      - `zip(*sentence)` 를 통해 sentence의 단어와 품사를 분리한다.
      - `append` 함수를 이용해 분리한 단어와 품사를 집어넣는다.
- 단어와 품사 리스트에 각각 단어 집합 생성  : `Tokenizer()`, `fit_on_texts`, 
- 단어와 품사 리스트에 각각 정수 인코딩 진행 : `texts_to_sequences`
- 단어와 품사 리스트에 각각 패딩 작업 : `pad_sequences`, `maxlen` 인자는 값이 같아야한다.
- 훈련 데이터와 테스트 데이터 분리 : `train_test_split`

#### 2) 양방향 LSTM으로 POS Tagger 만들기
- 레이블의 원-핫 인코딩 : 미리 진행하고 크로스 엔트로피 함수를 사용할 수 있지만 진행하지 않았다면 `sparse_categorical_crossentropy`를 선택할 수 있다.
- 모델 생성
     - 모델 생성 : `Sequential()`
     - 임베딩 층 생성 : `model.add(Embedding())`, 첫번째 인자는 입력 크기, 두번째 인자는 임베딩 층 출력의 크기
     - 은닉층 생성 : `model.add(Bidirectional(LSTM()))`, 첫번째 인자는 출력 크기, `return_sequences=True` 인자 추가
     - 출력 층 생성 : `model.add(TimeDistributed(Dense())))`, 첫번째 인자는 출력 크기, 활성화 함수는 소프트 맥스
     - 컴파일 : `model.compile`, 손실 함수는 `sparse_categorical_crossentropy`
- 단어와 품사 태깅 정보 리턴 : `tokenizer.index_word`

<br></br>

#### 1) 개체명 인식이란?
- 이름을 가진 개체를 인식하는 것
ex) "유진이는 8시에 밥을 먹었다." -> 유진 : 사람, 8시 : 시간, 밥 : 음식

#### 2) NLTK를 이용한 개체명 인식
- `from nltk import word_tokenize, pos_tag, ne_chunk`
- `word_tokenize` : 토큰화
- `pos_tag` : 품사 태깅
- `ne_chunk` : 개체명 인식, 앞에 품사 태깅이 진행되어야한다.

<br></br>

## 4. 개체명 인식의 BIO 표현 이해하기
- 개체명 인식기를 사용하는 것보다 목적에 맍는 데이터로 모델을 만드는 것이 좋다.

#### 1) BIO 표현
- B : 개체명이 시작하는 부분
- I : 개체명의 내부 부분
- O : 개체명이 아닌 부분
- BIO 뿐만 아니라 어떤 종류의 개체인지도 태깅이 된다.
ex) B-movie, I-movie, O

#### 2) 개체명 인식 데이터 이해
- CONLL2003 데이터 사용
- 형식 : 단어, 품사 태깅, 청크 태깅, 개체명 태깅

#### 3) 데이터 전처리
- 파일의 각 line을 읽는다.
- 공백을 기준으로 속석 분류 : `line.split`
- 정규표현식을 이용해 \n을 제거 : `re.sub`
- 단어들을 소문자로 저장 : `lower()`
- 단어와 개체명 태깅만 기록한다. : `append`
- line의 크기가 0이거나 줄바꿈 등이 일어나면 문장이 끝났다고 생각하여 전처리한 데이터를 넣는다.
  -> line 하나는 단어와 태깅 정보들이므로 여러 line을 읽고 한 문장이 한 샘플이 되도록 데이터를 만든다.
- 이전 챕터와 동일하게 단어와 품사태깅을 `zip` 함수를 이용해 분리한다.
- 단어 집합 생성 후 정수 인코딩, 패딩 작업 후 데이터 분리를 한다.
- 레이블에 해당하는 태깅 정보는 원-핫 인코딩을 한다.

#### 4) 양방향 LSTM으로 개체명 인식기 만들기
- 모델 생성 : `Sequential()`
- 임베딩 층 생성 : `model.add(Embedding())`, 입력 크기와 출력 크기를 지정해준다. 또한 `mask_zero=True`로 지정하면 패딩 작업 후 0이 많아질 경우 0은 연산에서 제외시킬 수 있다.
- 은닉층 생성 : `model.add(Bidirectional(LSTM()))`, `return_sequences=True` 로 지정해준다.
- 출력층 생성 : `model.add(TimeDistributed(Dense()))`, `TimeDistributed`는 LSTM을 다 대 다 구조로 수행할때 사용한다.
- 소프트 맥스 함수를 사용하고 손실함수로 `categorical_crossentropy`를 사용한다.

<br></br>

#### 1) 개체명 인식 데이터에 대한 이해와 전처리
- 데이터 링크  : https://www.kaggle.com/abhinavwalia95/entity-annotated-corpus
- `isnull().sum()` : 각 열마다 Null 값의 갯수를 보여준다.
- `fillna(method="ffill")` : Null 값을 가진 행은 바로 앞의 행의 값으로 Null을 채운다.
- 하나의 문장에 등장한 단어와 개체명을 묶는다. : https://wikidocs.net/147219 의 코드 확인
- `zip` 함수를 이용해 단어끼리, 개체명끼리 분리한다.
- 토큰화 ,정수 인코딩, 패딩, 데이터 분리 진행
     - 인덱스 1에 단어 "OOV"를 할당 : `Tokenizer(oov_token="OOV")`
     - 대문자를 유지 : `Tokenizer(lower=False)`
     - `fit_on_texts`, `texts_to_sequences`, `pad_sequences`, `train_test_split` 진행
- 레이블에는 원-핫 인코딩을 진행한다. : `to_categorical`

#### 2) 양방향 LSTM을 이용한 개체명 인식
- 이 전의 12.4에서의 양방향 LSTM으로 개체명 인식하는 코드와 비슷하다.

#### 3) F1-score
- 문제 : 상당 부분이 O인 데이터를 전체 O로 예측하면 다른 개체는 전부 찾지 못했음에도 높은 정확도를 얻는다.
- `pip install seqeval`
- 정밀도 : 특정 개체라고 예측한 경우 중 실제로 맞은 비율 -> (실제 P, 예측도 P)/{(실제 P, 예측 P)+(실제 N, 예측 P)}
- 재현률 : 실제 특정 개체 중 정답을 맞춘 비율 -> (실제 P, 예측 P)/{(실제 P, 예측 P) +(실제 P, 예측 N)}
- F1 score = 2*(정밀도 * 재현률)/(정밀도 + 재현률)
- 임포트 : `from seqeval.metrics import classification_report`
- 계산 : `classification_report([실제값], [예측값]))`
- precision : 정밀도, recall : 재현률, f1-score : F1-score

<br></br>

## 6. BiLSTM-CRF를 이용한 개체명 인식
- 기존 양방향 LSTM 모델에 CRF 층을 추가해 모델을 개선시킨다.

#### 1) CRF(Conditional Random Field)
- 문제 : I는 맨 앞에 올 수 없다. 하지만 양방향 LSTM만 고려하면 맨 앞의 값이 I이라고 예측할수도 있다.
- CRF 층 : 레이블 사이의 의존성을 고려한다.
- 활성화 함수를 지난 출력값이 CRF 층의 입력이 된다. -> 가장 높은 점수를 가지는 시퀀스를 예측한다.
- CRF 층이 훈련 데이터로 학습한 제약 사항
    - 문장의 첫 단어에 I가 오지않는다.
    - O-I 패턴은 나오지 않는다.
    - B-I-I 패턴은 개체명의 일관성을 유지한다. (개체명이 동일해야한다.)

#### 2) CRF 층 설치
- `pip install keras-crf`

#### 3) BiLSTM-CRF를 이용한 개체명 인식
- CRF 층 추가를 위해 Functional API를 사용한다.
- 모든 시점에 대해 개체명 레이블 갯수만큼의 선택지 중 하나를 예측해야하는 다중 클래스 분류
- Dropout 추가 : `TimeDistributed(Dropout(dropout_ratio))(model)`
- CRF 층 추가
    - 모델 생성  : `base = Model(inputs = input, outputs=output)`
    - 층 추가 : `CRFModel(base, tag_size)`, `tag_size`는 CRF 층에서 분류해야하는 선택지 갯수이다.
- keras-crf는 원-핫 인코딩 된 레이블을 지원하지 않으므로 정수 인코딩 된 레이블을 사용한다.
- 정수 인코딩 된 실제값과 예측값을 비교한다. : `classification_report`

<br></br>

#### 1) 문자 임베딩을 위한 전처리
- 문자 단위 정수 인코딩
- 문자 집합 생성 : 각 단어를 뽑고 각 단어를 `for` 문을 이용해 문자 단위로 분류하고 sorting 한다.
- 문자 집합으로부터 문자 -> 정수, 정수 -> 문자 딕셔너리를 만든다.

#### 2) BiLSTM-CNN을 이용한 개체명 인식
- 문자 임베딩이 임베딩 층을 통과하고 합성곱 층을 지나 풀링 층을 지난다.
- 그 결과와 단어 임베딩 벡터와 연결을 한다. : `concatenate([word, char])`
- 단어 임베딩 벡터와 연결하는 과정을 제외하고는 거의 비슷하다.

#### 3) BiLSTM-CNN-CRF
- 문자 임베딩을 사용하고 CRF 층까지 사용한다.
- 문자 결과 벡터와 단어 임베딩 벡터를 `concatenate`로 연결한다.
- `CRFModel`로 CRF 층을 넣어준다.

#### 4) BiLSTM-BiLSTM-CRF
- Conv1D를 수행하지 않고 단어 임베딩 결과와 문자 임베딩 결과를 BiLSTM을 거친 결과를 연결해서 LSTM을 수행한다.

---

- 참고
>> 딥 러닝을 이용한 자연어 처리 입문, 유원준, 안상준
      * 위 글 내용은 이 책의 내용을 공부하여 정리한 것입니다.
